dokka-subprojects/plugin-base/src/main/kotlin/org/jetbrains/dokka/base/translators/documentables/DefaultPageCreator.kt

```diff
@@ -643,128 +671,180 @@ public open class DefaultPageCreator(
+        } + extensions.groupBy {
+            NameAndIsExtension(it.name, isExtension = true)
+        }
+
+        val groups = grouped.entries
+            .sortedWith(compareBy(NameAndIsExtension.comparator) { it.key })
+            .map { (nameAndIsExtension, elements) ->
+                DivergentElementGroup(
+                    name = nameAndIsExtension.name,
+                    kind = when {
+                        nameAndIsExtension.isExtension -> ContentKind.Extensions
+                        else -> contentKind
+                    },
+                    contentTypeOverride = when {
+                        nameAndIsExtension.isExtension -> extensionContentType
+                        else -> null
+                    },
+                    elements = elements
+                )
+            }
+
         divergentBlock(
-            name,
-            list,
-            ContentKind.Properties,
-            extra = mainExtra
+            name = name,
+            kind = contentKind,
+            extra = mainExtra,
+            contentType = when {
+                // if only extensions
+                declarations.isEmpty() -> extensionContentType
+                else -> declarationContentType
+            },
+            groups = groups
         )
-
     }
-    private data class NameAndIsExtension(val name:String?, val isExtension: Boolean)
-
-    private fun groupAndSortDivergentCollection(collection: Collection<Documentable>): List<Map.Entry<NameAndIsExtension, List<Documentable>>> {
-        val groupKeyComparator: Comparator<Map.Entry<NameAndIsExtension, List<Documentable>>> =
-            compareBy<Map.Entry<NameAndIsExtension, List<Documentable>>, String?>(
-                nullsFirst(canonicalAlphabeticalOrder)
-            ) { it.key.name }
-                .thenBy { it.key.isExtension }
-
-       return collection
-            .groupBy {
-                NameAndIsExtension(
-                    it.name,
-                    it.isExtension()
-                )
-            } // This groupBy should probably use LocationProvider
-            // This hacks displaying actual typealias signatures along classlike ones
-            .mapValues { if (it.value.any { it is DClasslike }) it.value.filter { it !is DTypeAlias } else it.value }
-            .entries.sortedWith(groupKeyComparator)
+
+    private data class NameAndIsExtension(val name: String?, val isExtension: Boolean) {
+        companion object {
+            val comparator = compareBy(
+                comparator = nullsFirst(canonicalAlphabeticalOrder),
+                selector = NameAndIsExtension::name
+            ).thenBy(NameAndIsExtension::isExtension)
+        }
     }
 
-    protected open fun DocumentableContentBuilder.divergentBlock(
+    private class DivergentElementGroup(
+        val name: String?,
+        val kind: ContentKind,
+        val contentTypeOverride: BasicTabbedContentType?,
+        val elements: List<Documentable>
+    )
```

Comment:
`contentTypeOverride` looks suspicious... I wonder if it can be improved in any way

I think I understand why it exists - to be able to set `EXTENSION_FUNCTION` tab type for some of the children inside the `FUNCTION` block. 

And it seems to be a very specific corner case, as I can't imagine a scenario in which someone would want to override the type of tabbed content inside `BasicTabbedContentType.TYPE` - it would probably lead to a bug or undesired behaviour. So having it in one place like before seems more solid, i.e if `contentType == FUNCTION && group.elements.all { it.isExtension() } EXTENSION_FUNCTION else contentType`.

So what looks suspicious to me is that we're trying to generalize and encourage a corner case that maybe shouldn't even exist or should be refactored. It's not a major problem given that the class is private, we can just document it and hopefully it'll do until the next refactoring/feature/bugfix, but it would be nice to reduce the complexity here a little, so that you don't have to think about what exactly `contentTypeOverride` overrides, how to use it and in which scenarios

Do you maybe have some thoughts on this, or maybe you've tried different approaches and something didn't work / you didn't like?

As a side note: I honestly don't even remember why the differentiation between `FUNCTION` and `EXTENSION_FUNCTION` was needed - I all for removing it altogether if it makes things easier and everything works the same way as before. I hope we're not going to have anything similar with other types, like `CLASS` and `CLASS_INNER` or something.

Reply:
So, I've tried some different ways on how to improve code here, but after several tries I decided that it's not worth it in the end. While the current logic is not ideal, I've added some comments to improve its understanding.

Just for future readers, the most promising idea (at least I thought), both to clean up the code and simplify understanding of the logic was to add `BasicTabbedContentType` attribute no to the tab itself but to every row - in this case we clearly mark different rows with different type and so in HTML renderer depending on tab we filter by specific declarations (this also allows to easily do any splits, like showing only properties, or only functions, or other combinations if needed).
But it causes a lot of changes in final HTML, like some tags becomes just empty groups and they are skipped or new attributes are added. Also, this is needed only for functions/properties and make no sense for types (may be it's possible to extract `Types` to separate function).
So in the end, it will cause a lot of refactoring which could give us not a lot of profit (and also, could be refactored in future one more time)